# Project 2: Part 2
ST 590  
Summer 2022  
By Claudia Donahue and Collin Knezevich  

Use the sqlite database called northwind.sqlite. This database originally
comes from https://github.com/jpwhite3/northwind-SQLite3 but has been modified.
Your task is to write up a report about the employees. That is, you want to understand how much they sell,what products they are able to sell, how they’ve done across years, how they do in different regions, etc.  

You’ll want to study the database tables a bit to get an idea about the things you could investigate.  

You should provide summary statistics and graphs with corresponding interpretation. In the end, for each employee you should describe their main strength and something they can improve upon (backed up by your summary stats and graphs of course). Feel free to use SQL through pandas here to obtain the tables. You can do the joins, summaries, etc. through pandas if you’d like.

We reviewed the database and narrowed down a few things we wanted to investigate:  
- the value of sales for each employee  
- whether certain employees were more likely to have return customers  
- whether certain employees had higher revenue per sale  

First we connected to the database using the `sqlite3` package. 

In [1]:
import sqlite3

con = sqlite3.connect("northwind.sqlite")

Using `Pandas` dataframes, we started to investigate the contents of the database. First we grabbed a list of the tables present in the database. We also referenced a visualization of the database.  

In [35]:
# get a list of the tables in the db to make sure we have them right
import pandas as pd

tabs = pd.read_sql_query('SELECT name from sqlite_master where type= "table";', con) 
tabs

name
0             Categories
1        sqlite_sequence
2   CustomerCustomerDemo
3   CustomerDemographics
4              Customers
5              Employees
6    EmployeeTerritories
7          Order Details
8                 Orders
9               Products
10               Regions
11              Shippers
12             Suppliers
13           Territories

In [36]:
employees = pd.read_sql_query('SELECT * FROM Employees;', con) 
employees

EmployeeID   LastName FirstName                     Title TitleOfCourtesy  \
0           1    Davolio     Nancy      Sales Representative             Ms.   
1           2     Fuller    Andrew     Vice President, Sales             Dr.   
2           3  Leverling     Janet      Sales Representative             Ms.   
3           4    Peacock  Margaret      Sales Representative            Mrs.   
4           5   Buchanan    Steven             Sales Manager             Mr.   
5           6     Suyama   Michael      Sales Representative             Mr.   
6           7       King    Robert      Sales Representative             Mr.   
7           8   Callahan     Laura  Inside Sales Coordinator             Ms.   
8           9  Dodsworth      Anne      Sales Representative             Ms.   

    BirthDate    HireDate                         Address      City Region  \
0  1948-12-08  1992-05-01       507 - 20th Ave. E.Apt. 2A   Seattle     WA   
1  1952-02-19  1992-08-14              908 W. Capital Way    Tacoma     WA   
2  1963-08-30  1992-04-01              722 Moss Bay Blvd.  Kirkland     WA   
3  1937-09-19  1993-05-03            4110 Old Redmond Rd.   Redmond     WA   
4  1955-03-04  1993-10-17                 14 Garrett Hill    London   None   
5  1963-07-02  1993-10-17       Coventry House\nMiner Rd.    London   None   
6  1960-05-29  1994-01-02  Edgeham Hollow\nWinchester Way    London   None   
7  1958-01-09  1994-03-05           4726 - 11th Ave. N.E.   Seattle     WA   
8  1966-01-27  1994-11-15               7 Houndstooth Rd.    London   None   

  PostalCode Country       HomePhone Extension  \
0      98122     USA  (206) 555-9857      5467   
1      98401     USA  (206) 555-9482      3457   
2      98033     USA  (206) 555-3412      3355   
3      98052     USA  (206) 555-8122      5176   
4    SW1 8JR      UK   (71) 555-4848      3453   
5    EC2 7JR      UK   (71) 555-7773       428   
6    RG1 9SP      UK   (71) 555-5598       465   
7      98105     USA  (206) 555-1189      2344   
8    WG2 7LT      UK   (71) 555-4444       452   

                                               Photo  \
0  b'\xff\xd8\xff\xe0\x00\x10JFIF\x00\x01\x02\x00...   
1  b'\xff\xd8\xff\xe0\x00\x10JFIF\x00\x01\x02\x00...   
2  b'\xff\xd8\xff\xe0\x00\x10JFIF\x00\x01\x02\x00...   
3  b'\xff\xd8\xff\xe0\x00\x10JFIF\x00\x01\x02\x00...   
4  b'\xff\xd8\xff\xe0\x00\x10JFIF\x00\x01\x02\x00...   
5  b'\xff\xd8\xff\xe0\x00\x10JFIF\x00\x01\x02\x00...   
6  b'\xff\xd8\xff\xe0\x00\x10JFIF\x00\x01\x02\x00...   
7  b'\xff\xd8\xff\xe0\x00\x10JFIF\x00\x01\x02\x00...   
8  b'\xff\xd8\xff\xe0\x00\x10JFIF\x00\x01\x02\x00...   

                                               Notes  ReportsTo  \
0  Education includes a BA in psychology from Col...        2.0   
1  Andrew received his BTS commercial in 1974 and...        NaN   
2  Janet has a BS degree in chemistry from Boston...        2.0   
3  Margaret holds a BA in English literature from...        2.0   
4  Steven Buchanan graduated from St. Andrews Uni...        2.0   
5  Michael is a graduate of Sussex University (MA...        5.0   
6  Robert King served in the Peace Corps and trav...        5.0   
7  Laura received a BA in psychology from the Uni...        2.0   
8  Anne has a BA degree in English from St. Lawre...        5.0   

                                PhotoPath  
0    http://accweb/emmployees/davolio.bmp  
1     http://accweb/emmployees/fuller.bmp  
2  http://accweb/emmployees/leverling.bmp  
3    http://accweb/emmployees/peacock.bmp  
4   http://accweb/emmployees/buchanan.bmp  
5    http://accweb/emmployees/davolio.bmp  
6    http://accweb/emmployees/davolio.bmp  
7    http://accweb/emmployees/davolio.bmp  
8    http://accweb/emmployees/davolio.bmp

There are nine employees in the database. They include a Vice President of Sales, a Sales Manager, an Inside Sales Coordinator, and six Sales Representatives. They are located in either London or around Seattle. 

In [9]:
# I am not sure whether there is more data in this column
# Looks like it just trails off
employees['Notes']

0    Education includes a BA in psychology from Col...
1    Andrew received his BTS commercial in 1974 and...
2    Janet has a BS degree in chemistry from Boston...
3    Margaret holds a BA in English literature from...
4    Steven Buchanan graduated from St. Andrews Uni...
5    Michael is a graduate of Sussex University (MA...
6    Robert King served in the Peace Corps and trav...
7    Laura received a BA in psychology from the Uni...
8    Anne has a BA degree in English from St. Lawre...
Name: Notes, dtype: object

In [10]:
terr = pd.read_sql_query('SELECT * FROM Territories;', con) 
terr

TerritoryID                               TerritoryDescription  RegionID
0        01581  Westboro                                      ...         1
1        01730  Bedford                                       ...         1
2        01833  Georgetow                                     ...         1
3        02116  Boston                                        ...         1
4        02139  Cambridge                                     ...         1
5        02184  Braintree                                     ...         1
6        02903  Providence                                    ...         1
7        03049  Hollis                                        ...         3
8        03801  Portsmouth                                    ...         3
9        06897  Wilton                                        ...         1
10       07960  Morristown                                    ...         1
11       08837  Edison                                        ...         1
12       10019  New York                                      ...         1
13       10038  New York                                      ...         1
14       11747  Mellvile                                      ...         1
15       14450  Fairport                                      ...         1
16       19428  Philadelphia                                  ...         3
17       19713  Neward                                        ...         1
18       20852  Rockville                                     ...         1
19       27403  Greensboro                                    ...         1
20       27511  Cary                                          ...         1
21       29202  Columbia                                      ...         4
22       30346  Atlanta                                       ...         4
23       31406  Savannah                                      ...         4
24       32859  Orlando                                       ...         4
25       33607  Tampa                                         ...         4
26       40222  Louisville                                    ...         1
27       44122  Beachwood                                     ...         3
28       45839  Findlay                                       ...         3
29       48075  Southfield                                    ...         3
30       48084  Troy                                          ...         3
31       48304  Bloomfield Hills                              ...         3
32       53404  Racine                                        ...         3
33       55113  Roseville                                     ...         3
34       55439  Minneapolis                                   ...         3
35       60179  Hoffman Estates                               ...         2
36       60601  Chicago                                       ...         2
37       72716  Bentonville                                   ...         4
38       75234  Dallas                                        ...         4
39       78759  Austin                                        ...         4
40       80202  Denver                                        ...         2
41       80909  Colorado Springs                              ...         2
42       85014  Phoenix                                       ...         2
43       85251  Scottsdale                                    ...         2
44       90405  Santa Monica                                  ...         2
45       94025  Menlo Park                                    ...         2
46       94105  San Francisco                                 ...         2
47       95008  Campbell                                      ...         2
48       95054  Santa Clara                                   ...         2
49       95060  Santa Cruz                                    ...         2
50       98004  Bellevue                                      ...         2
51       98052  Redmond                            

In [11]:
empTerr = pd.read_sql_query('SELECT * FROM EmployeeTerritories;', con) 
empTerr

EmployeeID TerritoryID
0            1       06897
1            1       19713
2            2       01581
3            2       01730
4            2       01833
5            2       02116
6            2       02139
7            2       02184
8            2       40222
9            3       30346
10           3       31406
11           3       32859
12           3       33607
13           4       20852
14           4       27403
15           4       27511
16           5       02903
17           5       07960
18           5       08837
19           5       10019
20           5       10038
21           5       11747
22           5       14450
23           6       85014
24           6       85251
25           6       98004
26           6       98052
27           6       98104
28           7       60179
29           7       60601
30           7       80202
31           7       80909
32           7       90405
33           7       94025
34           7       94105
35           7       95008
36           7       95054
37           7       95060
38           8       19428
39           8       44122
40           8       45839
41           8       53404
42           9       03049
43           9       03801
44           9       48075
45           9       48084
46           9       48304
47           9       55113
48           9       55439

In [16]:
# The visualization of the db says 'Region' but it's 'Regions', turns out
region = pd.read_sql_query('SELECT * FROM Regions;', con) 
region

RegionID                                  RegionDescription
0         1  Eastern                                       ...
1         2  Westerns                                      ...
2         3  Northern                                      ...
3         4  Southern                                      ...

In [60]:
orders = pd.read_sql_query('SELECT * FROM Orders;', con) 

# Change ShipVia to a string data type
orders = orders.astype({"ShipVia": str}, errors='raise') 

orders

OrderID CustomerID  EmployeeID                OrderDate  \
0      10248      VINET           5  1996-07-04 00:00:00.000   
1      10249      TOMSP           6  1996-07-05 00:00:00.000   
2      10250      HANAR           4  1996-07-08 00:00:00.000   
3      10251      VICTE           3  1996-07-08 00:00:00.000   
4      10252      SUPRD           4  1996-07-09 00:00:00.000   
..       ...        ...         ...                      ...   
825    11073      PERIC           2  1998-05-05 00:00:00.000   
826    11074      SIMOB           7  1998-05-06 00:00:00.000   
827    11075      RICSU           8  1998-05-06 00:00:00.000   
828    11076      BONAP           4  1998-05-06 00:00:00.000   
829    11077      RATTC           1  1998-05-06 00:00:00.000   

                RequiredDate              ShippedDate ShipVia  Freight  \
0    1996-08-01 00:00:00.000  1996-07-16 00:00:00.000       3    32.38   
1    1996-08-16 00:00:00.000  1996-07-10 00:00:00.000       1    11.61   
2    1996-08-05 00:00:00.000  1996-07-12 00:00:00.000       2    65.83   
3    1996-08-05 00:00:00.000  1996-07-15 00:00:00.000       1    41.34   
4    1996-08-06 00:00:00.000  1996-07-11 00:00:00.000       2    51.30   
..                       ...                      ...     ...      ...   
825  1998-06-02 00:00:00.000                     None       2    24.95   
826  1998-06-03 00:00:00.000                     None       2    18.44   
827  1998-06-03 00:00:00.000                     None       2     6.19   
828  1998-06-03 00:00:00.000                     None       2    38.28   
829  1998-06-03 00:00:00.000                     None       2     8.53   

                       ShipName               ShipAddress        ShipCity  \
0     Vins et alcools Chevalier        59 rue de l-Abbaye           Reims   
1           Toms Spezialitäten             Luisenstr. 48        Münster   
2                 Hanari Carnes          Rua do Paço, 67  Rio de Janeiro   
3          Victuailles en stock        2, rue du Commerce            Lyon   
4            Suprêmes délices      Boulevard Tirou, 255       Charleroi   
..                          ...                       ...             ...   
825  Pericles Comidas clásicas  Calle Dr. Jorge Cash 321    México D.F.   
826               Simons bistro             Vinbæltet 34       Kobenhavn   
827          Richter Supermarkt               Starenweg 5         Genève   
828                    Bon app-      12, rue des Bouchers       Marseille   
829  Rattlesnake Canyon Grocery           2817 Milton Dr.     Albuquerque   

    ShipRegion ShipPostalCode  ShipCountry  
0         None          51100       France  
1         None          44087      Germany  
2           RJ      05454-876       Brazil  
3         None          69004       France  
4         None         B-6000      Belgium  
..         ...            ...          ...  
825       None           5033       Mexico  
826       None           1734      Denmark  
827       None           1204  Switzerland  
828       None          13008       France  
829         NM          87110          USA  

[830 rows x 14 columns]

In the orders table, we can note the timeframe of the sales was from mid 1996 to mid 1998. We can group by employee over those two years to get an idea of how many sales each employee made. 

In [61]:
orders.groupby('EmployeeID').count().sort_values(by=['OrderID'], ascending=False)

OrderID  CustomerID  OrderDate  RequiredDate  ShippedDate  \
EmployeeID                                                              
4               156         156        156           156          151   
3               127         127        127           127          127   
1               123         123        123           123          120   
8               104         104        104           104          100   
2                96          96         96            96           93   
7                72          72         72            72           69   
6                67          67         67            67           65   
9                43          43         43            43           42   
5                42          42         42            42           42   

            ShipVia  Freight  ShipName  ShipAddress  ShipCity  ShipRegion  \
EmployeeID                                                                  
4               156      156       156          156       156          62   
3               127      127       127          127       127          56   
1               123      123       123          123       123          50   
8               104      104       104          104       104          42   
2                96       96        96           96        96          31   
7                72       72        72           72        72          22   
6                67       67        67           67        67          32   
9                43       43        43           43        43          14   
5                42       42        42           42        42          14   

            ShipPostalCode  ShipCountry  
EmployeeID                               
4                      155          156  
3                      122          127  
1                      122          123  
8                      103          104  
2                       93           96  
7                       70           72  
6                       64           67  
9                       40           43  
5                       42           42

We noticed there was no column annotating the value of each order, so we had to keep searching to find a way to compare sales. We thought that the total value of sales per employee would be more interesting than the quantity of sales. We found that price information in the Order Details table: 

In [62]:
orderDetails = pd.read_sql_query('SELECT * FROM "Order Details";', con) 

# change ProductID to a string data type
orderDetails = orderDetails.astype({"ProductID": str}, errors='raise') 

orderDetails

OrderID ProductID  UnitPrice  Quantity  Discount
0       10248        11      14.00        12      0.00
1       10248        42       9.80        10      0.00
2       10248        72      34.80         5      0.00
3       10249        14      18.60         9      0.00
4       10249        51      42.40        40      0.00
...       ...       ...        ...       ...       ...
2150    11077        64      33.25         2      0.03
2151    11077        66      17.00         1      0.00
2152    11077        73      15.00         2      0.01
2153    11077        75       7.75         4      0.00
2154    11077        77      13.00         2      0.00

[2155 rows x 5 columns]

So we were able to merge the Orders table with the Order Details table and add up the total revenue per employee over the time period considered. 

In [63]:
# merge
orders1 = orders.merge(orderDetails, on=['OrderID'])

# change OrderID to a string data type
orders1 = orders1.astype({"OrderID": str}, errors='raise') 

# create Revenue column
orders1["Revenue"] = ((orders1["UnitPrice"] * orders1["Quantity"]) * (1 - orders1["Discount"])).round(2)

# Rank employees by total revenue
orders1.groupby('EmployeeID').sum().sort_values(by=['Revenue'], ascending=False)

Freight  UnitPrice  Quantity  Discount    Revenue
EmployeeID                                                    
4           37261.55   11187.29      9798     25.75  232890.87
3           35783.62    8748.42      7852     15.95  202812.82
1           28118.25    8732.44      7812     16.99  192107.59
2           25924.53    6913.41      6055     10.55  166537.76
8           23304.42    6239.25      5913     14.60  126862.27
7           21667.67    4717.51      4654     12.95  124568.22
9           10215.69    3380.20      2670      7.30   77308.06
6           11123.09    3834.96      3527      9.15   73913.14
5           13907.28    2747.43      3036      7.80   68792.29

The range of revenue per employee was pretty big. From $68,792 (assuming we are working with US dollars) up to $232,891.  

One interesting aspect of this chart is the use of discounts by different employees. We wanted to explore that further.  
We also were interested in the revenue per sale for each employee. 